### PR TITLE
add a Win32 askpass

### PR DESCRIPTION
**Pull request by rimrul**
It's currently as bare-bones as it gets, but seems to be functional.

grafik

The prompt does get wrapped into multiple lines if it gets too long. The code is obviously heavily based on the [sample code for line edits](https://docs.microsoft.com/en-us/windows/win32/controls/use-a-single-line--edit-control) and slightly on `git-askyesno`.

I intend to add the Git for Windows icon on the right (similar to the GCM `askpass`) at some point in the near future, but that'll probably happen after v2.33.0 is out.

The name isn't set in stone yet, but I think shipping two significantly different `git-askpass.exe` that are more or less intended for the same purpose might lead to confusion.

Changes since v1:
* rename to `git-askpass`
* includ the changes to the Makefile that I forgot to add
* switch from `DialogBoxW()` to `DialogBoxParamW()`. This allows passing the prompt directly and gives us easier access to the result of the Dialog
* return non-zero on cancel (alligns behaviour closer with OpenSSH `ssh-askpass`)
* drop the changing of the default button from the sample code (alligns behaviour closer with OpenSSH `ssh-askpass`)
* pass prompt directly
* prefer `%ls` over `%S`
* also show single arg prompts 
* fixed checksums that where off due to wrong EOLs
* prepared `env.sh` to also handle `MSYSTEM=ARM64`
* added the new `askpass` to `env.sh`

### CODE REVIEW

**Review by dscho — COMMENTED** [review]
Very nice!

File: git-extra/git-win32-askpass.c [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+					SendDlgItemMessageW(hDlg,
+									   IDE_PASSWORDEDIT,
+									   EM_GETLINE,
+									   (WPARAM) 0,       /* line 0 */
+									   (LPARAM) lpszPassword);
+
+					/* Null-terminate the string. */
+					lpszPassword[cchPassword] = 0;
+
+					wprintf(L"%S\n", lpszPassword);
```
[the remark this answers is not in the document]
Comment (rimrul) [reply git-extra/git-win32-askpass.c:64]: Yes, I can do that.

File: git-extra/git-win32-askpass.c [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+					EndDialog(hDlg, TRUE);
+					free(lpszPassword);
+					return TRUE;
+
+				case IDCANCEL:
+					EndDialog(hDlg, TRUE);
```
[the remark this answers is not in the document]
Comment (rimrul) [reply git-extra/git-win32-askpass.c:71]: We'd need to call `DialogBoxParamW()` directly to actually recieve a return value, but that would also allow us to pass the prompt directly. Could be a good idea.

File: git-extra/PKGBUILD [hunk 3]
```diff
@@ -52,7 +52,10 @@ source=('inputrc'
         'update-via-pacman.bat'
         'git-credential-helper-selector.c'
         'git-credential-helper-selector.manifest'
-        'git-credential-helper-selector.rc')
+        'git-credential-helper-selector.rc'
+        'git-win32-askpass.c'
```
[the remark this answers is not in the document]
Comment (rimrul) [reply git-extra/PKGBUILD:56]: Would that cause a new name conflict with microsoft/git-credential-manager-core#385 or obsolete that issue?

File: git-extra/git-win32-askpass.c [hunk 4]
```diff
@@ -0,0 +1,118 @@
+#define WIN32_LEAN_AND_MEAN
+#include <windows.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include "git-win32-askpass.h"
+
+wchar_t *prompt = NULL;
```
[the remark this answers is not in the document]
Comment (rimrul) [reply git-extra/git-win32-askpass.c:7]: I'll move it back into `wmain()` and switch from `DialogBoxW()` to `DialogBoxParamW()`

**Comment by rimrul** [comment]
I think I've addressed all concerns. I've also added the new `askpass` to `env.sh` and prepared `env.sh` to handle `ARM64`.

**Comment by dscho** [comment]
I modified this a little, locally. This is the range-diff:

```diff
1:  7e0a028098 ! 1:  da9ba5d86b git-extra: include an simple Win32 `git askpass`
    @@ Commit message
         Signed-off-by: Matthias Aßhauer <mha1993@live.de>
     
      ## git-extra/Makefile ##
    +@@ git-extra/Makefile: CXXFLAGS ?= -Wall
    + all: $(BUILDDIR)/create-shortcut.exe $(BUILDDIR)/WhoUses.exe \
    + 	$(BUILDDIR)/blocked-file-util.exe $(BUILDDIR)/proxy-lookup.exe \
    + 	$(BUILDDIR)/git-askyesno.exe \
    ++	$(BUILDDIR)/git-askpass.exe \
    + 	$(BUILDDIR)/git-credential-helper-selector.exe
    + 
    + $(BUILDDIR)/create-shortcut.exe: $(BUILDDIR)/create-shortcut.o
     @@ git-extra/Makefile: $(BUILDDIR)/git-credential-helper-selector.exe: \
      	$(BUILDDIR)/git-credential-helper-selector.o \
      	$(BUILDDIR)/git-credential-helper-selector.res
    @@ git-extra/git-askpass.c (new)
     +		case WM_INITDIALOG:
     +			/* Display the prompt */
     +			SetDlgItemTextW(hDlg, IDC_PROMPT, (wchar_t *) lParam);
    -+			return TRUE; 
    ++			if (GetDlgCtrlID((HWND) wParam) != IDE_PASSWORDEDIT) {
    ++				SetFocus(GetDlgItem(hDlg, IDE_PASSWORDEDIT));
    ++				return FALSE;
    ++			}
    ++			return TRUE;
     +		case WM_COMMAND:
     +			switch(wParam)
     +			{
     +				case IDOK:
     +					/* Get number of characters. */
     +					cchPassword = (WORD) SendDlgItemMessage(hDlg,
    -+															IDE_PASSWORDEDIT,
    -+															EM_LINELENGTH,
    -+															(WPARAM) 0,
    -+															(LPARAM) 0);
    ++										IDE_PASSWORDEDIT,
    ++										EM_LINELENGTH,
    ++										(WPARAM) 0,
    ++										(LPARAM) 0);
     +
     +					lpszPassword = (wchar_t *) malloc(sizeof(wchar_t) * (cchPassword + 1));
     +					if (!lpszPassword) {
    @@ git-extra/git-askpass.c (new)
     +
     +					/* Get the characters. */
     +					SendDlgItemMessageW(hDlg,
    -+									   IDE_PASSWORDEDIT,
    -+									   EM_GETLINE,
    -+									   (WPARAM) 0,       /* line 0 */
    -+									   (LPARAM) lpszPassword);
    ++							    IDE_PASSWORDEDIT,
    ++							    EM_GETLINE,
    ++							    (WPARAM) 0,       /* line 0 */
    ++							    (LPARAM) lpszPassword);
     +
     +					/* Null-terminate the string. */
     +					lpszPassword[cchPassword] = 0;
    @@ git-extra/git-askpass.c (new)
     +{
     +	INT_PTR res;
     +	wchar_t *prompt = NULL;
    -+	
    ++
     +	if (argc < 2) {
     +		MessageBoxW(NULL, L"Usage: git askpass <prompt>", L"Error!", MB_OK);
     +		return 1;
    @@ git-extra/git-askpass.c (new)
     +	free(prompt);
     +	return !res;
     +}
    - \ No newline at end of file
     
      ## git-extra/git-askpass.h (new) ##
     @@
2:  1976019925 = 2:  dc55377e79 git-extra: set `DISPLAY` and `SSH_ASKPASS` for ARM64
3:  350810e4a9 = 3:  c292bffd28 env.sh: prefer git-askpass.exe over git-gui--askpass
```

Summary:

1. `git-askpass.exe` is now built upon `make`
2. The edit box gets the focus
3. Whitespace issues were fixed via `git rebase --whitespace=fix`
4. I fixed some indentation manually

**Comment by dscho** [comment]
`git-extra` is being built and uploaded here: https://dev.azure.com/Git-for-Windows/git/_build/results?buildId=82787&view=results

**Comment by dscho** [comment]
Thank you so much @rimrul!!!

File: git-extra/PKGBUILD (hunk 3 above)
[the remark this answers is not in the document]
Comment (dscho) [reply git-extra/PKGBUILD]: I think it will obsolete https://github.com/microsoft/Git-Credential-Manager-Core/issues/385

**Comment by dscho** [comment]
Oh, and once this is built, we will have to fix `mingit/release.sh`, I think: it would probably see a conflict otherwise because `libexec/git-core/*` is moved to `bin/` in MinGit.

**Comment by rimrul** [comment]
> `libexec/git-core/*` is moved to `bin/` in MinGit.

Is it? I'm looking at the contents of MinGit-2.32.0.2-64-bit.zip and `git-askpass.exe` (and about 46 other files) is in `libexec/git-core` not `bin/`

**Comment by dscho** [comment]
Hrm. I thought I had moved them all into `bin/`. It's actually quite important.

**Comment by dscho** [comment]
> 
> 
> Hrm. I thought I had moved them all into `bin/`. It's actually quite important.

Turns out that I only moved the `.exe` files from `libexec/git-core/` to `bin/` that originate from the `mingw-w64-git` package: https://github.com/git-for-windows/build-extra/blob/36a1cbd1502a062384d1fce715c9c3d2fde36b34/mingit/release.sh#L126-L140

So I think we're actually good here.

**Pull request merged**